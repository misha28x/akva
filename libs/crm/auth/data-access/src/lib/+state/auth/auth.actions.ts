import { createAction, props } from '@ngrx/store';

import { Credentials, User } from '../../models/auth.model';

export const login = createAction(
  '[Login Page] Login',
  props<{ credentials: Credentials }>()
);

export const loginSuccess = createAction(
  '[Auth Api] Login Success',
  props<{ user: User }>()
);
export const loginFailure = createAction(
  '[Auth Api] Login Failure',
  props<{ error: any }>()
);

export const logout = createAction('[Auth] Logout');
