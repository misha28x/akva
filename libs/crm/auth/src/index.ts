export * from './lib/crm-auth.module';
export * from './lib/state/auth-facade.service';
