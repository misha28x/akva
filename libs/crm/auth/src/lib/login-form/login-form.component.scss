@import 'variables';

:host {
  .form {
    &__label {
      font-size: $x-large-fs;
      letter-spacing: -0.5x;
      line-height: 1;

      margin: 0 0 $margin-lg;
    }

    &__section {
      margin: $margin-lg 0;
    }

    &__button {
      width: 100%;
    }
  }
}
