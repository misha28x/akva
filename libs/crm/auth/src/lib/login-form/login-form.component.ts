import {
  ChangeDetectionStrategy,
  EventEmitter,
  Component,
  Output,
  Input,
} from '@angular/core';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';

import { Credentials } from '@akva/shared/auth-models';

@Component({
  selector: 'akva-login-form',
  templateUrl: './login-form.component.html',
  styleUrls: ['./login-form.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class LoginFormComponent {
  @Input() error: string;
  @Input() loading: boolean;

  @Output() submitted = new EventEmitter<Credentials>();

  loginForm: FormGroup = this.createLoginForm();

  constructor(private fb: FormBuilder) {}

  onSubmit() {
    this.loginForm.markAllAsTouched();
    if (this.loginForm.invalid) {
      return;
    }

    this.submitted.emit(this.getCredentials());
  }

  private getCredentials(): Credentials {
    return this.loginForm.value as Credentials;
  }

  private createLoginForm() {
    return this.fb.group({
      login: ['', [Validators.required]],
      password: ['', [Validators.required]],
    });
  }
}
