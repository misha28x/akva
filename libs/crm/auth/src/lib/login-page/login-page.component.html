<section class="content">
  <div class="logo content__logo">
    <img
      class="logo__img"
      alt="Логотип Аквастандарт"
      src="./assets/logo-white.png"
    />
  </div>

  <div class="content__card">
    <akva-login-form
      [loading]="loading"
      (submitted)="onSubmit($event)"
    ></akva-login-form>
  </div>
</section>
