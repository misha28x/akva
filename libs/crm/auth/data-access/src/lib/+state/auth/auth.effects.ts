import { Injectable } from '@angular/core';
import { Router } from '@angular/router';

import { Actions, createEffect, ofType } from '@ngrx/effects';

import { of } from 'rxjs';
import { catchError, exhaustMap, map, tap } from 'rxjs/operators';

import { login, loginFailure, loginSuccess, logout } from './auth.actions';
import { AuthService } from '../../services/auth.service';
import { JwtService } from '../../services/jwt.service';

@Injectable()
export class AuthEffects {
  constructor(
    private router: Router,
    private actions$: Actions,
    private jwtService: JwtService,
    private authService: AuthService
  ) {}

  login$ = createEffect(() =>
    this.actions$.pipe(
      ofType(login),
      exhaustMap(({ credentials }) =>
        this.authService.login(credentials).pipe(
          tap(({ token }) => this.jwtService.storeToken(token)),
          map(({ user }) => loginSuccess({ user })),
          catchError(({ errorMsg }) => {
            return of(loginFailure({ error: errorMsg }));
          })
        )
      )
    )
  );

  logoutRedirect$ = createEffect(
    () =>
      this.actions$.pipe(
        ofType(logout),
        tap(() => this.router.navigate(['/login']))
      ),
    {
      dispatch: false,
    }
  );

  loginSuccessRedirect$ = createEffect(
    () =>
      this.actions$.pipe(
        ofType(loginSuccess),
        tap(() => this.router.navigate(['/']))
      ),
    { dispatch: false }
  );
}
