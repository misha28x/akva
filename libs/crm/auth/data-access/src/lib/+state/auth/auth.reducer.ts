import { createReducer, on, Action } from '@ngrx/store';

import * as AuthActions from './auth.actions';
import { User } from '../../models/auth.model';

export const AUTH_FEATURE_KEY = 'auth';

export interface State {
  user?: User | null;
  error?: string | null; // last known error (if any)
  pending?: boolean;
}

export interface AuthPartialState {
  readonly [AUTH_FEATURE_KEY]: State;
}

export const initialState: State = {
  pending: false,
};

const authReducer = createReducer(
  initialState,
  on(AuthActions.login, (state) => ({ ...state, pending: true })),
  on(AuthActions.loginSuccess, (state, { user }) => ({
    ...state,
    user,
    pending: false,
  })),
  on(AuthActions.loginFailure, (state, { error }) => ({
    ...state,
    error,
    pending: false,
  }))
);

export function reducer(state: State | undefined, action: Action) {
  return authReducer(state, action);
}
