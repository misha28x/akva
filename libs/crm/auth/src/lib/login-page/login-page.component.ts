import { Component } from '@angular/core';

import { AuthService } from '../services/auth.service';
import { Credentials } from '@akva/shared/auth-models';

@Component({
  selector: 'akva-login',
  templateUrl: './login-page.component.html',
  styleUrls: ['./login-page.component.scss'],
})
export class LoginPageComponent {
  get loading() {
    return this.status === 'loading';
  }

  private status: 'idle' | 'loading' | 'submitted' | 'error' = 'idle';

  constructor(private auth: AuthService) {}

  onSubmit(credentials: Credentials) {
    this.status = 'loading';
    this.authorize(credentials).subscribe(() => (this.status = 'idle'));
  }

  private authorize(credentials: Credentials) {
    return this.auth.login(credentials);
  }
}
