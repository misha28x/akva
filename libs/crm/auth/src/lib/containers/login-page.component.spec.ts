import {
  async,
  ComponentFixture,
  fakeAsync,
  TestBed,
  tick,
} from '@angular/core/testing';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { By } from '@angular/platform-browser';
import { of, asyncScheduler } from 'rxjs';

import { API_PROVIDER } from '@akva/shared/config';

import { LoginPageComponent } from './login-page.component';
import { LoginFormComponent } from '../components/login-form.component';
import { AuthFacadeService } from '@akva/crm/auth';

import { Credentials } from '@akva/shared/auth-models';

const authFacadeFactory = (): Partial<AuthFacadeService> => ({
  login: jest.fn(),
  loading$: of(false, asyncScheduler),
  error$: of('', asyncScheduler),
});

describe('LoginPageComponent', () => {
  let credentials: Credentials;
  let component: LoginPageComponent;
  let fixture: ComponentFixture<LoginPageComponent>;
  let service;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [LoginPageComponent, LoginFormComponent],
      imports: [FormsModule, ReactiveFormsModule],
      providers: [
        API_PROVIDER,
        { provide: AuthFacadeService, useValue: authFacadeFactory() },
      ],
      schemas: [NO_ERRORS_SCHEMA],
    }).compileComponents();
  }));

  beforeEach(() => {
    credentials = { password: 'admin', username: 'admin' };
    fixture = TestBed.createComponent(LoginPageComponent);
    component = fixture.componentInstance;
    service = TestBed.inject(AuthFacadeService);
  });

  afterEach(() => {
    service.login.mockRestore();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should authorize when credentials is submitted', fakeAsync(() => {
    const form = fixture.debugElement.query(By.css('akva-login-form'))
      .componentInstance as LoginFormComponent;

    form.submitted.emit(credentials);
    tick();

    expect(service.login).toBeCalledWith(credentials);
  }));
});
