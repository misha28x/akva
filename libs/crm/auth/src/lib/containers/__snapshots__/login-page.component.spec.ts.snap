// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPageComponent should compile 1`] = `
<akva-login
  error$={[Function Store]}
  loading$={[Function Store]}
  store={[Function MockStore]}
>
  <section
    class="content"
  >
    <div
      class="logo content__logo"
    >
      <img
        alt="Логотип Аквастандарт"
        class="logo__img"
        src="./assets/logo-white.png"
      />
    </div>
    <div
      class="content__card"
    >
      <akva-login-form />
    </div>
  </section>
</akva-login>
`;
